File: iwapp_sandp/events/stock_reconciliation.py
import frappe

def validate(doc, method):
    # if doc.items:
        # for i in doc.items:
        #     if i.custom_model_id:
        #         i.custom_brand = frappe.db.get_value("Item Model ID", i.custom_model_id, "brand")
        #         i.custom_description = f"{i.item_code} - {i.custom_brand} {i.custom_model_id}"
    if doc.items:
        for item in doc.items:
            if item.custom_from_model_id == 0:
                if item.custom_model_id:
                    brand_and_spec = frappe.db.get_value('Item Model ID', item.custom_model_id, ["brand", "specification"])
                    if brand_and_spec:
                        brand = brand_and_spec[0] if brand_and_spec[0] else ""
                        specification = brand_and_spec[1] if brand_and_spec[1] else ""
                        if specification:
                            item.custom_description = f"{item.item_code} - {brand} {item.custom_model_id} - {specification}"
                        else:
                            item.custom_description = f"{item.item_code} - {brand} {item.custom_model_id}"
                        item.custom_brand = brand
                        item.custom_from_model_id = 1

def on_submit(doc, method):
    if doc.items:
        for item in doc.items:
            if item.serial_no:
                frappe.db.set_value("Item Model ID", item.custom_model_id, "serial_no", item.serial_no)
                # Replace commas with newline characters and split the string into a list and Remove spaces from each element in the list
                data_list = [i.strip() for i in item.serial_no.replace(',', '\n').split('\n') if i]
                for i in data_list:
                    frappe.db.set_value("Serial No", i, {"custom_model_id" : item.custom_model_id, "brand":item.custom_brand, "serial_no_details" : item.custom_description, "custom_update_model_id":1})

File: iwapp_sandp/public/js/stock_reconciliation.js
frappe.ui.form.on('Stock Reconciliation', {
    refresh: function (frm) {
        if (frm.doc.docstatus != 1) {
            frm.add_custom_button(__('Fetch Model Id'), function () {
                if (frm.doc.items) {
                    $.each(frm.doc.items, function (idx, item) {
                        if (item.serial_no) {
                            let dataString = item.serial_no
                            let dataList = dataString.replace(/,/g, '\n').split('\n').map(item => item.trim()).filter(Boolean);
                            $.each(dataList, function (idx, serial_no) {
                                frappe.db.get_value("Serial No", serial_no, 'custom_model_id')
                                    .then(r => {
                                        if (r.message.custom_model_id) {
                                            frappe.model.set_value(item.doctype, item.name, "custom_model_id", r.message.custom_model_id)
                                        }
                                    })
                            })
                        }
                    })
                }
            })
        }
        frm.fields_dict['items'].grid.get_field('custom_model_id').get_query = function (doc, cdt, cdn) {
            var child = locals[cdt][cdn];
            return {
                filters: {
                    'item': child.item_code
                }
            };
        };
    }
});

frappe.ui.form.on('Stock Reconciliation Item', {
    custom_model_id: function (frm, cdt, cdn) {
        var item = locals[cdt][cdn];
        frappe.model.set_value(cdt, cdn, "custom_description", "")
        if (item.custom_model_id) {
            frappe.db.get_value('Item Model ID', item.custom_model_id, ["brand", "specification"]).then(response => {
                const brand_and_spec = response.message;
                if (brand_and_spec) {
                    const brand = brand_and_spec.brand || "";
                    const specification = brand_and_spec.specification || "";
                    if (specification) {
                        const description = `${item.item_name}, ${brand} - ${item.custom_model_id}, ${specification}`;
                        frappe.model.set_value(cdt, cdn, "custom_description", description)
                    } else {
                        const description = `${item.item_name}, ${brand} - ${item.custom_model_id}`;
                        frappe.model.set_value(cdt, cdn, "custom_description", description)
                    }
                }
            });
        }
    },
    item_code: function (frm, cdt, cdn) {
        var item = locals[cdt][cdn];
        if (item.custom_model_id) {
            frappe.model.set_value(cdt, cdn, "custom_model_id", "")
        }
    }
});